processor/spanmetricsprocessor/processor.go

```diff
@@ -140,16 +191,215 @@ func (p *processorImp) ConsumeTraces(ctx context.Context, traces pdata.Traces) e
 
 // buildMetrics collects the computed raw metrics data, builds the metrics object and
 // writes the raw metrics data into the metrics object.
-func (p *processorImp) buildMetrics() *pdata.Metrics {
-	// TODO: Add implementation
+func (p *processorImp) buildMetrics() (*pdata.Metrics, error) {
+	ilm := pdata.NewInstrumentationLibraryMetrics()
+	ilm.InstrumentationLibrary().SetName("spanmetricsprocessor")
+
+	p.lock.RLock()
+	if err := p.collectCallMetrics(&ilm); err != nil {
+		return nil, err
+	}
+	if err := p.collectLatencyMetrics(&ilm); err != nil {
+		return nil, err
+	}
+	p.lock.RUnlock()
+
+	rm := pdata.NewResourceMetrics()
+	ilms := rm.InstrumentationLibraryMetrics()
+	ilms.Append(ilm)
+
 	m := pdata.NewMetrics()
-	return &m
+	m.ResourceMetrics().Append(rm)
+	return &m, nil
+}
+
+// collectLatencyMetrics collects the raw latency metrics, writing the data
+// into the given instrumentation library metrics.
+func (p *processorImp) collectLatencyMetrics(ilm *pdata.InstrumentationLibraryMetrics) error {
+	for key := range p.latencyCount {
+		dpLatency := pdata.NewIntHistogramDataPoint()
+		dpLatency.SetTimestamp(pdata.TimeToUnixNano(time.Now()))
+		dpLatency.SetExplicitBounds(p.latencyBounds)
+		dpLatency.SetBucketCounts(p.latencyBucketCounts[key])
+		dpLatency.SetCount(p.latencyCount[key])
+		dpLatency.SetSum(int64(p.latencySum[key]))
+		kmap := make(map[string]string)
+		if err := p.jsonSerder.Unmarshal([]byte(key), &kmap); err != nil {
+			return err
+		}
+		dpLatency.LabelsMap().InitFromMap(kmap)
+
+		mLatency := pdata.NewMetric()
+		mLatency.SetDataType(pdata.MetricDataTypeIntHistogram)
+		mLatency.SetName("latency")
+		mLatency.IntHistogram().DataPoints().Append(dpLatency)
+		mLatency.IntHistogram().SetAggregationTemporality(pdata.AggregationTemporalityCumulative)
+		ilm.Metrics().Append(mLatency)
+	}
+	return nil
+}
+
+// collectCallMetrics collects the raw call count metrics, writing the data
+// into the given instrumentation library metrics.
+func (p *processorImp) collectCallMetrics(ilm *pdata.InstrumentationLibraryMetrics) error {
+	for key := range p.callSum {
+		dpCalls := pdata.NewIntDataPoint()
+		dpCalls.SetTimestamp(pdata.TimeToUnixNano(time.Now()))
```

Comment:
Do we also need to set the start time? See https://github.com/open-telemetry/opentelemetry-proto/blob/b9f9b7e6d0ecbf00784a02e1b17fd5cc00ee7d73/opentelemetry/proto/metrics/v1/metrics.proto#L265

Reply:
From what I understand from the link you provided, setting a start time a requirement when periodically resetting cumulative metrics, is that right?

Reply:
I think so, see my previous reply.